### PR TITLE
Multi step plugin archive

### PR DESCRIPTION
**Pull request by Biont**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Refactoring

**What is the current behavior?** (You can also link to an open issue here)
`build-plugin-archive.yml` uses a single php version for bundling dependencies and executing build tools.
WP's committment to supporting php down to 7.0 makes it increasingly cumbersome to support **"one"** *php platform requirement* for production AND development/building.

We have been accepting this limitation in order to streamline our packaging across most of our projects - and it's been a very solid base. But the fact remains that the php ecosystem is moving on while the php7 support requirement for WP packages is not going anywhere. Since this gap is only getting wider, we need to find a better approach

**What is the new behavior (if this is a feature change)?**
The logic of the workflow has been split into 3 parts:

1. **Production dependencies** - checkout the repo and `composer install` using the ancient PHP version
2. **Build steps** - execute all relevant build tools using a much more modern php version. We also no longer expect these to be present in dev dependencies: If we find a config, we install the tool globally. 
3. **Archive creation** - add final touches and create the resulting archive

Note: There is no strict technical requirement for step 3 to be in a separate job. It adds a little overhead, but I think at this point, it adds a nice separation of concerns in that it keeps a few environment variables and data handling out of the other steps.

The individual jobs each produce their own "interim" artifact which is then picked up by the next one. This makes it easy to spot and debug where something went wrong:

screenshot_2024-03-20-120520

---

This change frees us from having to use EOL versions of common tools and also paves the way for modernization of our own tools, for example [dropping PHP 7 in Composer Asset Compiler](https://github.com/inpsyde/composer-asset-compiler/pull/23). It also allows us to introduce third-party tools like *php-scoper* with ease since we are no longer subject to the php version constraints of individual projects.

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

**Other information**:

### CODE REVIEW

File: .github/workflows/build-plugin-archive.yml [hunk 1]
```diff
@@ -118,38 +104,82 @@ jobs:
...
+          hasTranslateConfig(){
+            jq '.extra | has("wp-translation-downloader")' < composer.json >/dev/null 2>&1
+            local EXIT=$?
+            echo "$EXIT"
+          }
           hasDep(){
             composer show -i -N -D --strict "${1}" >/dev/null 2>&1
             local EXIT=$?
             echo "$EXIT"
           }
```
Comment (tyrann0us) [inline .github/workflows/build-plugin-archive.yml:170]: `hasDep()` is no longer used.
```suggestion
```

File: .github/workflows/build-plugin-archive.yml [hunk 2]
```diff
@@ -118,38 +104,82 @@ jobs:
...
+        run: |
+          if [ "${{ inputs.PACKAGE_MANAGER }}" == 'npm' ] && { [ -f "${GITHUB_WORKSPACE}/package-lock.json" ] || [ -f "${GITHUB_WORKSPACE}/npm-shrinkwrap.json" ]; }; then
+            echo "NODE_CACHE_MODE=npm" >> $GITHUB_ENV
+          elif [ "${{ inputs.PACKAGE_MANAGER }}" == 'yarn' ] && [ -f "${GITHUB_WORKSPACE}/yarn.lock" ]; then
+            echo "NODE_CACHE_MODE=yarn" >> $GITHUB_ENV
+          else
+            echo "No lock files found or unknown package manager"
+          fi
 
       - name: Check for package.json
```
Comment (tyrann0us) [inline .github/workflows/build-plugin-archive.yml:184]: I know this is existing code, but why do we need this step? We already check for the presence of the Composer Asset Compiler configuration; there's no valid reason for it to be present in a package that does _not_ have a `package.json` file.

Also, the `check-for-package` step ID just confused me while reading the code. It is always a package; what we do here is to check if it is an _npm_ package or rather if it has a `package.json` file. If you decide to keep this step, can we at least improve the ID?

File: .github/workflows/build-plugin-archive.yml [hunk 3]
```diff
@@ -168,32 +198,77 @@ jobs:
...
           cache: ${{ env.NODE_CACHE_MODE }}
 
+      - name: Install Composer Asset Compiler
+        if: steps.composer-tools.outputs.assets-compiler == '0' && steps.check-for-package.outputs.has_package == 'true'
+        run: |
+          composer global config --no-plugins --no-interaction allow-plugins.inpsyde/composer-assets-compiler true
+          composer global require inpsyde/composer-assets-compiler
       - name: Run Composer Asset Compiler
         if: steps.composer-tools.outputs.assets-compiler == '0' && steps.check-for-package.outputs.has_package == 'true'
         run: composer compile-assets ${{ inputs.COMPILE_ASSETS_ARGS }}
```
Comment (tyrann0us) [inline .github/workflows/build-plugin-archive.yml:208]: Can we merge this into one step?
```suggestion
      - name: Install and run Composer Asset Compiler
        if: steps.composer-tools.outputs.assets-compiler == '0'
        run: |
          composer global config --no-plugins --no-interaction allow-plugins.inpsyde/composer-assets-compiler true
          composer global require inpsyde/composer-assets-compiler
          composer compile-assets ${{ inputs.COMPILE_ASSETS_ARGS }}
```

Note: I already removed the `has_package` check (see comment above).

File: .github/workflows/build-plugin-archive.yml [hunk 4]
```diff
@@ -168,32 +198,77 @@ jobs:
...
 
+      - name: Install WordPress Translation Downloader
+        if: steps.composer-tools.outputs.translation-downloader == '0'
+        run: |
+          composer global config --no-plugins --no-interaction allow-plugins.composer/installers true
+          composer global config --no-plugins --no-interaction allow-plugins.inpsyde/wp-translation-downloader true
+          composer global require inpsyde/wp-translation-downloader
       - name: Run WordPress Translation Downloader
         if: steps.composer-tools.outputs.translation-downloader == '0'
         run: composer wp-translation-downloader:download
```
Comment (tyrann0us) [inline .github/workflows/build-plugin-archive.yml:218]: Same as above.
```suggestion
      - name: Install and run WordPress Translation Downloader
        if: steps.composer-tools.outputs.translation-downloader == '0'
        run: |
          composer global config --no-plugins --no-interaction allow-plugins.composer/installers true
          composer global config --no-plugins --no-interaction allow-plugins.inpsyde/wp-translation-downloader true
          composer global require inpsyde/wp-translation-downloader
          composer wp-translation-downloader:download
```

File: .github/workflows/build-plugin-archive.yml [hunk 5]
```diff
@@ -25,7 +25,7 @@ on:
         type: string
       COMPOSER_ARGS:
         description: Set of arguments passed to Composer.
-        default: '--no-dev --no-scripts --prefer-dist --optimize-autoloader'
+        default: '--no-dev --prefer-dist --optimize-autoloader'
         required: false
         type: string
       PHP_VERSION:
```
Comment (tyrann0us) [inline .github/workflows/build-plugin-archive.yml:31]: The description needs to be updated, I guess.

File: .github/workflows/build-plugin-archive.yml [hunk 6]
```diff
@@ -168,32 +198,77 @@ jobs:
...
       - name: Add commit hash to plugin header
         run: 'sed -Ei "s/SHA: .*/SHA: ${GIT_SHA}/g" ${{ inputs.PLUGIN_MAIN_FILE }}'
 
       - name: Set plugin version header
         run: 'sed -Ei "s/Version: .*/Version: ${PLUGIN_VERSION}/g" ${{ inputs.PLUGIN_MAIN_FILE }}'
 
+      - name: Set up PHP
+        uses: shivammathur/setup-php@v2
+        with:
+          php-version: '8.3'
```
Comment (tyrann0us) [inline .github/workflows/build-plugin-archive.yml:263]: Let's please not hardcode a PHP version here. I understand you want to use a different version than `${{ inputs.PHP_VERSION }}`, but in this case, please create an additional input (which ideally would default to PHP 8.0 to be aligned with all other PHP version inputs; see #108).

**Review by tyrann0us — CHANGES_REQUESTED** [review]

File: .github/workflows/build-plugin-archive.yml (hunk 1 above)
Comment (Biont) [reply .github/workflows/build-plugin-archive.yml:170]: Yes but I am letting go of this awesome piece of scripting with a heavy heart, just so you know

File: .github/workflows/build-plugin-archive.yml (hunk 3 above)
Comment (Biont) [reply .github/workflows/build-plugin-archive.yml:203]: Okay makes sense if you prefer it that way

File: .github/workflows/build-plugin-archive.yml (hunk 6 above)
Comment (Biont) [reply .github/workflows/build-plugin-archive.yml:254]: Ha! I was actually just trying to see if the `tools:` section would make life easier by setting up `php-scoper` in the `PATH` as opposed to my previous hamfisted approach of finding the executable.
I even recall thinking "maybe this should be a new input" but then forgot. I'll add this

File: .github/workflows/build-plugin-archive.yml (hunk 5 above)
Comment (Biont) [reply .github/workflows/build-plugin-archive.yml:31]: Done

File: .github/workflows/build-plugin-archive.yml (hunk 6 above)
Comment (Biont) [reply .github/workflows/build-plugin-archive.yml]: Done

File: .github/workflows/build-plugin-archive.yml (hunk 2 above)
Comment (Biont) [reply .github/workflows/build-plugin-archive.yml]: I removed it. Thanks for pointing this out

File: docs/archive-creation.md [hunk 7]
```diff
@@ -46,7 +46,8 @@ jobs:
 | `NPM_REGISTRY_DOMAIN` | `"https://npm.pkg.github.com/"`                               | Domain of the private npm registry                                                             |
 | `PACKAGE_MANAGER`     | `"yarn"`                                                      | Package manager with which the dependencies should be installed (`npm` or `yarn`)              |
 | `COMPOSER_ARGS`       | `'--no-dev --no-scripts --prefer-dist --optimize-autoloader'` | Set of arguments passed to Composer                                                            |
```
Comment (tyrann0us) [inline docs/archive-creation.md:48]:
```suggestion
| `COMPOSER_ARGS`       | `'--no-dev --no-scripts --prefer-dist --optimize-autoloader'` | Set of arguments passed to Composer when gathering production dependencies |
```
Please ensure to align the table.

**Review by tyrann0us — APPROVED** [review]
Thanks for incorporating the changes so quickly. LGTM. Left a minor comment that doesn't stop me from approving. 💪🏽

File: docs/archive-creation.md (hunk 7 above)
Comment (Biont) [reply docs/archive-creation.md:48]: I did reformat using the PHPStorm inspection, but yeah I forgot to update the description for `COMPOSER_ARGS`. Thanks for that!

**Pull request merged**